ls1intum/Artemis
# Submit StudentExam after Exam ended

### Checklist

### Motivation and Context


When the working time is over, the latest student exam state should be transferred to the server. To distribute the amount of requests, the request is made once the student accepts the end terms of the exam

### Description

Added fields to entities: 
- Student Exam - boolean submitted (marks whether an exam is submitted or not)
- Exam - gracePeriod integer (time for accepting the terms)

Server:
- add endpoint to submit student exam
- add service method to submit student exam
- generating student exam: set submitted to false for new student exams

Client:
- put timer into own component
- display timer also on exam-participation-cover with grace period
- add call to server once 
- show exam title on end exam-participation-cover


### Steps for Testing


1. Create an exam and participate (working for modeling, text and quiz exercises)
2. Make last minute changes to an exercise (should not be synchronized with the server via the regular synchronization mechanisms (manual save, 30s timer or exercise change)). 
3. When the exam finishes, the user sees the end page where he/she has to accept the end terms. The student gets a grace period of 3minutes in which he/she can accept the end terms of the exam
4. When the terms are accepted, the last version of the student exam is sent to the server and the student exam is marked as submitted. This includes the last minute changes from the previous step (working for modeling, text and quiz exercises) 
5. the student can then see the summary
6. refresh the summary page to see if your submissions are correctly synchronized (the aforementioned last minute changes - working for modeling, text and quiz exercises) 
7. once accepted, the end terms should not be displayed again (e.g. after page reload)

#### Student does not accept terms
1. if the student did not accept the terms, the summary is not displayed
2. if the end terms are not accepted until the exam is over (after grace period is 0 - timer in the end page, see screenshot), the summary cannot be displayed and the end terms can no longer be accepted


### Screenshots


student does not accept terms
![image](https://user-images.githubusercontent.com/22955066/87257448-48e28080-c49b-11ea-8d95-4c0eb2171850.png)
